hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3AccessGrantConfiguration.java

```diff
@@ -0,0 +1,108 @@
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import static org.apache.hadoop.fs.s3a.Constants.AWS_S3_ACCESS_GRANTS_ENABLED;
+
+
+/**
+ * Test S3 Access Grants configurations.
+ */
+public class TestS3AccessGrantConfiguration extends AbstractHadoopTestBase {
+    /**
+     * This credential provider will be attached to any client
+     * that has been configured with the S3 Access Grants plugin.
+     * {@link software.amazon.awssdk.s3accessgrants.plugin.S3AccessGrantsPlugin}.
+     */
+    public static final String S3_ACCESS_GRANTS_EXPECTED_CREDENTIAL_PROVIDER_CLASS =
+            S3AccessGrantsIdentityProvider.class.getName();
+
+    @Test
+    public void testS3AccessGrantsEnabled() throws IOException, URISyntaxException {
+        // Feature is explicitly enabled
+        AwsClient s3AsyncClient = getAwsClient(createConfig(true), true);
+        Assert.assertEquals(
+                S3_ACCESS_GRANTS_EXPECTED_CREDENTIAL_PROVIDER_CLASS,
+                getCredentialProviderName(s3AsyncClient));
+
+        AwsClient s3Client = getAwsClient(createConfig(true), false);
+        Assert.assertEquals(
+                S3_ACCESS_GRANTS_EXPECTED_CREDENTIAL_PROVIDER_CLASS,
+                getCredentialProviderName(s3Client));
+    }
+
+    @Test
+    public void testS3AccessGrantsDisabled() throws IOException, URISyntaxException {
+        // Disabled by default
+        AwsClient s3AsyncDefaultClient = getAwsClient(new Configuration(), true);
+        Assert.assertNotEquals(
+                S3_ACCESS_GRANTS_EXPECTED_CREDENTIAL_PROVIDER_CLASS,
+                getCredentialProviderName(s3AsyncDefaultClient));
+
+        AwsClient s3DefaultClient = getAwsClient(new Configuration(), true);
+        Assert.assertNotEquals(
+                S3_ACCESS_GRANTS_EXPECTED_CREDENTIAL_PROVIDER_CLASS,
+                getCredentialProviderName(s3DefaultClient));
+
+        // Disabled if explicitly set
+        AwsClient s3AsyncExplicitlyDisabledClient = getAwsClient(createConfig(false), true);
+        Assert.assertNotEquals(
+                S3_ACCESS_GRANTS_EXPECTED_CREDENTIAL_PROVIDER_CLASS,
+                getCredentialProviderName(s3AsyncExplicitlyDisabledClient));
+
+        AwsClient s3ExplicitlyDisabledClient = getAwsClient(createConfig(false), true);
+        Assert.assertNotEquals(
+                S3_ACCESS_GRANTS_EXPECTED_CREDENTIAL_PROVIDER_CLASS,
+                getCredentialProviderName(s3ExplicitlyDisabledClient));
+    }
+
+    private Configuration createConfig(boolean s3agEnabled) {
+        Configuration conf = new Configuration();
+        conf.setBoolean(AWS_S3_ACCESS_GRANTS_ENABLED, s3agEnabled);
+        return conf;
+    }
+
+    private String getCredentialProviderName(AwsClient awsClient) {
+        return awsClient.serviceClientConfiguration().credentialsProvider().getClass().getName();
+    }
+
+    private <BuilderT extends S3BaseClientBuilder<BuilderT, ClientT>, ClientT> AwsClient
```

Comment:
since you're returning an instance of `AwsClient` , i don't think you need generics here?

Reply:
Yup, changed.